File: trabalho1.py
import re
import itertools
import math
import cPickle as pickle
import numpy as np
import cv2

data_dir = sys.argv[1]
stopwords_file = sys.argv[2]

STRIPCHARS = '-_\'"'
WORD_REGEX = re.compile('[A-Z]{2,}(?![a-z])|[A-Z][a-z]+(?=[A-Z])|[\'a-zA-Z\-]+')
def split_words_from_email(email, stopwords):
    words = map(lambda w: w.lower().strip(STRIPCHARS),
                WORD_REGEX.findall(email))

    return [w for w in words if len(w) >= 3 and w not in stopwords]

# read emails and remove stop words
with open(stopwords_file) as f:
    stopwords = set(str.strip(word) for word in f)

all_words = set()
word_count_in_email = {} # (word, email) -> how many times word is in email
emails_with_word_count = {} # word -> how many emails contain the word

email_files = glob.glob(os.path.join(data_dir, '*.txt'))
email_names = map(os.path.basename, email_files)

for i, ef in enumerate(email_files):
    email = email_names[i]

    with open(ef) as f:
        sys.stderr.write('\rReading email {} of {}... '.format(i + 1, len(email_names)))

        words_in_email = split_words_from_email(f.read(), stopwords)
        unique_words = set(words_in_email)
        all_words |= unique_words

        for w in words_in_email:
            word_count_in_email[w, email] = word_count_in_email.setdefault((w, email), 0) + 1

        for w in unique_words:
            emails_with_word_count[w] = emails_with_word_count.setdefault(w, 0) + 1

print 'Done!'

# calculate tf (term frequency) and idf (inverse document frequency) for
# each word and email, and pick the words we'll use as features

idf = {}
feature_words = []
for w in all_words:
    ratio = len(email_files) / float(emails_with_word_count[w])
    idf[w] = math.log(ratio)

tf = word_count_in_email

print 'Using {} unique words.'.format(len(all_words))

# calculate the term vector for each email

term_vectors = []

out_queue = []
OUT_QUEUE_MAXSIZE = 1024
with open('term_vectors', 'w') as tvf:
    for i, email in enumerate(email_names):
        sys.stderr.write('\rCalculating tf {} of {}... '.format(i + 1, len(email_names)))

        tvec = tuple(idf[w] * tf.get((w, email), 0) for w in all_words)

        # normalize to unit length
        norm = np.linalg.norm(tvec)
        tvec = tuple(e / norm for e in tvec)

        term_vectors.append(tvec)

        serialized = repr((email, tvec))
        assert eval(serialized) == (email, tvec)

        out_queue.append(serialized)
        if len(out_queue) == OUT_QUEUE_MAXSIZE:
            while out_queue:
                tvf.write(out_queue.pop(0) + '\n')

    while out_queue:
        tvf.write(out_queue.pop(0) + '\n')

term_vectors = np.array(term_vectors, dtype = 'float32')
distortion, clusters, means = cv2.kmeans(
        term_vectors, K = 23, criteria = (cv2.TERM_CRITERIA_MAX_ITER, 1000, 0),
        attempts = 5, flags = cv2.KMEANS_RANDOM_CENTERS)

print distortion

File: trabalho1.r
library("Matrix")
library("RWeka")

EMAILS_DIR = "messages-dist"
STOPWORDS = "stopwords/english"
LOG = paste("logs/Log -", date())

log.message <- function(message)
{
    cat(paste(message, "\n", sep = ""), file = LOG, append = TRUE)
}

word.cache.size <- 0
make.words.from.email <- function(email.name, email.file, stopwords)
{
    cache.key <- paste("cached.words.for", email.name, sep = ".")
    if( exists(cache.key) )
    {
        return(get(cache.key))
    }

    email.lines <- readLines(email.file)
    tokens <- Map(tolower, WordTokenizer(email.lines))
    sanitized <- Map(function(t) { gsub("[\\'\"_*-<>=|%{}~^]", "", t) }, tokens)
    stemmed <- LovinsStemmer(sanitized[!sanitized %in% stopwords])
    words <- stemmed[nchar(stemmed) > 2]

    # cache results
    assign("word.cache.size", word.cache.size + 1, envir = .GlobalEnv)
    assign(cache.key, words, envir = .GlobalEnv)

    return(words)
}

compute.full.dictionary <- function(email.files, stopwords)
{
    # all unique words
    all.words <- Reduce(
        function(all.words.so.far, ef)
        {
            email <- basename(ef)
            words.in.email <- make.words.from.email(email, ef, stopwords)
            return(union(all.words.so.far, words.in.email))
        }, email.files, init = c())

    return(all.words)
}

compute.feature.words <- function(raw.dictionary, count.emails.with.word)
{
    too.infrequent <- count.emails.with.word[count.emails.with.word <= 2]
    log.message(paste("Found", length(too.infrequent), "infrequent words"))

    feature.words <- setdiff(raw.dictionary, names(too.infrequent))
    return(feature.words)
}

compute.feature.vectors <- function()
{
    log.message("Starting up!")

    stopwords <- readLines(STOPWORDS)
    email.files <- Sys.glob(file.path(EMAILS_DIR, "*.txt"))

    raw.dictionary <- compute.full.dictionary(email.files, stopwords)
    log.message(paste("Found", length(raw.dictionary), "unique words"))
    log.message(paste("Cached words for", word.cache.size, "emails"))

    # calculate how many emails contain a word
    count.emails.with.word <- rep(0, length(raw.dictionary))
    names(count.emails.with.word) <- raw.dictionary

    progress <- 0
    for( ef in email.files )
    {
        email <- basename(ef)
        words.in.email <- make.words.from.email(email, ef, stopwords)

        # XXX this already eliminates duplicates for some reason
        count.emails.with.word[words.in.email] <-
            count.emails.with.word[words.in.email] + 1

        progress <- progress + 1
        if( progress %% 100 == 0 )
        {
            log.message(paste("count.emails.with.word for", progress, "emails"))
        }
    }

    # decide which words will end up being features
    feature.words <- compute.feature.words(raw.dictionary, count.emails.with.word)
    log.message(paste("Ended up with", length(feature.words), "words"))
    remove(raw.dictionary)

    # calculate the idf (inverse document frequency) for each feature word
    count.emails.with.word <- count.emails.with.word[feature.words]
    idf <- log(length(email.files) / count.emails.with.word)
    remove(count.emails.with.word)

    # compute tf, the term frequency (number of times each word appears
    # in each email)
    tf <- Matrix(0, nrow = length(email.files), ncol = length(feature.words))
    rownames(tf) <- Map(basename, email.files)
    colnames(tf) <- feature.words

    progress <- 0
    for( ef in email.files )
    {
        email <- basename(ef)
        words.in.email <- make.words.from.email(email, ef, stopwords)

        count.feature.words.table <-
            table(words.in.email[words.in.email %in% feature.words])
        feature.words.in.email <- rownames(count.feature.words.table)

        count.feature.words <- as.vector(count.feature.words.table)

        tf[email,feature.words.in.email] <-
            tf[email,feature.words.in.email] + count.feature.words

        progress <- progress + 1
        if( progress %% 100  == 0 )
        {
            log.message(paste("tf for", progress, "emails"))
        }
    }

    log.message("Done computing idf and tf!")

    fv.not.normalized <- tf * idf
    fv <- fv.not.normalized / sqrt(rowSums(fv.not.normalized ^ 2))

    # export results to global env
    assign("fv", fv, envir = .GlobalEnv)
    assign("feature.words", feature.words, envir = .GlobalEnv)
    assign("tf", tf, envir = .GlobalEnv)
    assign("idf", idf, envir = .GlobalEnv)
}

do.kmeans <- function(k.values)
{
    log.message("Beginning kmeans")
    kmeans.results <- sapply(k.values,
        function(k)
        {
            log.message(paste("kmeans for", k, "clusters"))
            kmeans(fv, k)
        })

    # export results to global env
    assign("kmeans.results", kmeans.results, envir = .GlobalEnv)
}

main <- function()
{
    compute.feature.vectors()
}

main()
